// @mancho.devs/authorizer/dist/__test__/index.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const testData_1 = require("./testData");
const testSigner_1 = require("./testSigner");
describe('Signer', () => {
    const requestData = {
        body: { payment: 'payment', amount: 100 },
        headers: {
            Host: 'api.paymentsgateway.averspay.kg',
            'x-api-timestamp': '1636026186643',
            'x-api-key': 'h8z1TDStxu5YY2YuN8jUa9hpzIVbfkLT7kPiPiYj',
            'x-api-header': 'header value',
            'X-Forwarded-For': '54.187.127.20, 15.158.4.100',
            'X-Forwarded-Port': '443',
            'X-Forwarded-Proto': 'https',
        },
        httpMethod: 'POST',
        path: '/services',
        queryStringParameters: {
            from: '0',
            size: '10',
            test: undefined,
            best: 'тест дата',
        },
    };
    it('should successfully get http method', () => {
        const signer = new testSigner_1.SignerForTest(requestData);
        expect(signer.getHttpMethod()).toEqual('post');
    });
    it('should fail to get headers data when header has an invalid value', () => {
        const request = {
            ...requestData,
            headers: {
                ...requestData.headers,
                'x-api-test': undefined,
            },
        };
        const signer = new testSigner_1.SignerForTest(request);
        expect(() => {
            signer.getHeadersData();
        }).toThrow(`Header 'x-api-test' contains invalid value`);
    });
    it('should fait to get headers data when empty headers provided', () => {
        const request = {
            ...requestData,
            headers: {},
        };
        const signer = new testSigner_1.SignerForTest(request);
        expect(() => {
            signer.getHeadersData();
        }).toThrow(Error(`Header 'Host' is required`));
    });
    it('should successfully get headers data when normal header provided', () => {
        const signer = new testSigner_1.SignerForTest(requestData);
        expect(signer.getHeadersData()).toEqual('host:api.paymentsgateway.averspay.kg&x-api-header:header value&x-api-key:h8z1TDStxu5YY2YuN8jUa9hpzIVbfkLT7kPiPiYj&x-api-timestamp:1636026186643');
    });
    it('should successfully get query string params data when query string params not provided', () => {
        const request = {
            ...requestData,
            queryStringParameters: null,
        };
        const signer = new testSigner_1.SignerForTest(request);
        expect(signer.getQueryStringParamsData()).toEqual('');
    });
    it('should successfully get query string params data when query string params provided', () => {
        const signer = new testSigner_1.SignerForTest(requestData);
        expect(signer.getQueryStringParamsData()).toEqual('best=%D1%82%D0%B5%D1%81%D1%82%20%D0%B4%D0%B0%D1%82%D0%B0&from=0&size=10&test=');
    });
    it('should successfully get encoded query string params data', () => {
        const data = {
            ...requestData,
            queryStringParameters: {
                from: '0',
                size: '10',
                test: undefined,
                best: '%D1%82%D0%B5%D1%81%D1%82%20%D0%B4%D0%B0%D1%82%D0%B0',
            },
        };
        const signer = new testSigner_1.SignerForTest(data);
        expect(signer.getQueryStringParamsData()).toEqual('best=%D1%82%D0%B5%D1%81%D1%82%20%D0%B4%D0%B0%D1%82%D0%B0&from=0&size=10&test=');
    });
    it('should successfully get simple path', () => {
        const data = { ...requestData, path: '/services/жаны-cервис' };
        const signer = new testSigner_1.SignerForTest(data);
        expect(signer.getPath()).toEqual('/services/жаны-cервис');
    });
    it('should successfully get encoded path', () => {
        const data = { ...requestData, path: '/services/%D0%B6%D0%B0%D0%BD%D1%8B-c%D0%B5%D1%80%D0%B2%D0%B8%D1%81' };
        const signer = new testSigner_1.SignerForTest(data);
        expect(signer.getPath()).toEqual('/services/жаны-cервис');
    });
    it('should successfully get json body', () => {
        const signer = new testSigner_1.SignerForTest(requestData);
        expect(signer.getJsonBody()).toEqual('{"amount":100,"payment":"payment"}');
    });
    it('should successfully get empty string when there is no body', () => {
        const request = {
            ...requestData,
            body: null,
        };
        const signer = new testSigner_1.SignerForTest(request);
        expect(signer.getJsonBody()).toEqual('');
    });
    it('should successfully get data', () => {
        const signer = new testSigner_1.SignerForTest(requestData);
        expect(signer.getData()).toEqual('post\n/services\nhost:api.paymentsgateway.averspay.kg&x-api-header:header value&x-api-key:h8z1TDStxu5YY2YuN8jUa9hpzIVbfkLT7kPiPiYj&x-api-timestamp:1636026186643\nbest=%D1%82%D0%B5%D1%81%D1%82%20%D0%B4%D0%B0%D1%82%D0%B0&from=0&size=10&test=\n{"amount":100,"payment":"payment"}');
    });
    it('should successfully sign', async () => {
        const signer = new testSigner_1.SignerForTest(requestData);
        const signature = await signer.sign(testData_1.privateKey);
        expect(signature).toEqual('TiQCuWaV1WE/VDsbYKn6O0B2diji6MyZI6zjC8Q9lEdnc6KkxURnot1i874fw8q5cyBpLXO6T7dH70VpC11pT1vlrZDZe+PzGnYe27pRqwxU6KcohG5iYp5eeUjQHNaJHL/7zkJdCRu6nIj0z84xbLYMYbPBfTHPPp+viwnGqEdR4wIcjVm18Op3WKgOj5zTv2HB4ATNi31nERYN2R3/ecn+CgK8tIf6Ox3azhNJat3oIQT6Gk10wvAROLsNFKm82Px3CeT/lXO1d8UeeTMNGe8mvo7POGUrH4UJhjsa1myvpNyKeW1vF1kuSv8bFcoJfkXbiZ51gHGxpoL8MmYhlA==');
    });
    it('should successfully verify', async () => {
        const signer = new testSigner_1.SignerForTest(requestData);
        const signature = await signer.sign(testData_1.privateKey);
        const isValid = signer.verify(testData_1.publicKey, signature);
        expect(isValid).toBeTruthy();
    });
    it('should fail to verify with incorrect signature', async () => {
        const signer = new testSigner_1.SignerForTest(requestData);
        const isValid = await signer.verify(testData_1.publicKey, testData_1.incorrectSignText);
        expect(isValid).toBeFalsy();
    });
    it('should successfully match sign with simple path and encoded path', async () => {
        const encodedPathdata = {
            ...requestData,
            path: '/services/%D0%B6%D0%B0%D0%BD%D1%8B-c%D0%B5%D1%80%D0%B2%D0%B8%D1%81',
        };
        const signer1 = new testSigner_1.SignerForTest(encodedPathdata);
        const simplePathData = { ...requestData, path: '/services/жаны-cервис' };
        const signer2 = new testSigner_1.SignerForTest(simplePathData);
        expect(await signer1.sign(testData_1.privateKey)).toEqual(await signer2.sign(testData_1.privateKey));
    });
});
