// @mancho.devs/authorizer/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Signer = void 0;
/* External dependencies */
const crypto_1 = require("crypto");
const node_jose_1 = require("node-jose");
class Signer {
    requestData;
    constructor(requestData) {
        console.log('request data:', requestData);
        this.requestData = requestData;
    }
    getHttpMethod() {
        return this.requestData.httpMethod.toLowerCase();
    }
    getPath() {
        return decodeURI(this.requestData.path ?? '');
    }
    getHeadersData() {
        const { headers } = this.requestData;
        const { Host } = headers;
        if (!Host) {
            throw new Error(`Header 'Host' is required`);
        }
        const headerData = `host:${Host.toString()}`;
        const sortedHeadersKeys = Object.keys(headers)
            .filter((key) => key.toLowerCase().startsWith('x-api-'))
            .sort();
        const headersData = sortedHeadersKeys.map((key) => {
            const value = headers[key];
            if (typeof value === 'undefined' || value === null || typeof value.toString !== 'function') {
                throw new Error(`Header '${key}' contains invalid value`);
            }
            const headerData = key.toLowerCase() + ':' + value.toString();
            return headerData;
        });
        return [headerData, ...headersData].join('&');
    }
    getQueryStringParamsData() {
        const queryParams = this.requestData.queryStringParameters ?? {};
        const sortedQueryParamKeys = Object.keys(queryParams).sort();
        const queryParamsData = sortedQueryParamKeys.map((key) => {
            const value = queryParams[key] ?? '';
            const queryParamData = encodeURI(decodeURI(key)) + '=' + encodeURI(decodeURI(value));
            return queryParamData;
        });
        return queryParamsData.join('&');
    }
    getJsonBody() {
        let body = '';
        if (this.requestData.body) {
            // sort object by key
            const sortedBody = Object.entries(this.requestData.body)
                .sort((a, b) => a[0].localeCompare(b[0]))
                .reduce((result, [key, value]) => {
                result[key] = value;
                return result;
            }, {});
            body = JSON.stringify(sortedBody);
        }
        return body;
    }
    getData() {
        const parts = [];
        const queryString = this.getQueryStringParamsData();
        parts.push(this.getHttpMethod());
        parts.push(this.getPath());
        parts.push(this.getHeadersData());
        if (queryString) {
            parts.push(queryString);
        }
        parts.push(this.getJsonBody());
        return parts.join('\n');
    }
    async sign(privateKey) {
        const keyStore = node_jose_1.JWK.createKeyStore();
        const rsaKey = await keyStore.add(privateKey, 'pem');
        const signer = (0, crypto_1.createSign)('SHA256');
        const data = this.getData();
        console.log('Data to be signed:', data);
        signer.update(data);
        const signature = signer.sign(rsaKey.toPEM(true), 'base64');
        return signature;
    }
    async verify(publicKey, signature) {
        const keyStore = node_jose_1.JWK.createKeyStore();
        const rsaKey = await keyStore.add(publicKey, 'pem');
        const signer = (0, crypto_1.createVerify)('SHA256');
        const data = this.getData();
        console.log('Data to be verified:', data);
        signer.update(data);
        const isValid = signer.verify(rsaKey.toPEM(), signature, 'base64');
        return isValid;
    }
}
exports.Signer = Signer;
